### ISSUE TITLE
Object gets modified before kubemod can modify it , leading to error

### ISSUE THREAD

**Issue by rohitkumar1369** [kubemod/kubemod#119]
I am using kubemod-operator in my clusters with a simple node's synthetic reference injection rule mentioned here : https://github.com/kubemod/kubemod#synthetic-references 
I am using the one in example that allows injection of a node's region and zone into pod's metadata. I am running it on several large scale clusters ( nodes > 500 ) and pods per namespace > 500 .
However I am running intermittently into the following error message : 
`{"level":"error","ts":"2023-09-13 09:00:50.881Z","logger":"podbinding-webhook","msg":"failed to inject annotation ref.kubemod.io/nodename into pod","request uid":"01272605-5c73-4ecd-93ed-8945c7c49cd2","namespace":"follow-service","resource":"pods/<pod_name>l","operation":"CREATE","podNamespace":"<namespace>","podName":"<pod_name>","error":"Operation cannot be fulfilled on pods \"<pod_name>\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\ngithub.com/kubemod/kubemod/core.(*PodBindingWebhookHandler).Handle\n\t/workspace/core/podbindingwebhook.go:95\nsigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/webhook/admission/webhook.go:135\nsigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/webhook/admission/http.go:87\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:76\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2462\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2916\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1966"}`

This is very intermittent and has no relation to a specific type of node. 
Does kubemod have a retry mechanism that can force a retry on failed requests? I need the node's region and zone into pod's metadata as it is crucial for a zone-aware proxy that I'm trying to run as a sidecar.

**Comment by vassilvk** [comment]
Hi @rohitkumar1369, I see what the problem is.

It would make sense to retry on dirty-write failures - my concern here is that such retries would typically require a back-off, which will ultimately slow down the response by the webhook, and that's not a good thing, because one needs to get back from a webhook as soon as possible.

I am thinking of revamping this logic with an async solution where the pod binding webhook triggers an async process which takes care of applying the annotation and can backoff as much as it needs.

**Comment by rohitkumar1369** [comment]
Hi @vassilvk 
If you are planning to revamp this logic to an async one, then can I assume that for injecting node labels ( synthetic references ) don't need this podbinding webhook call ?

**Comment by vassilvk** [comment]
The podbinding webhook is the foundation of KubeMod's ability to get node-related data and inject into pods.
When a pod is created in Kubernetes, initially the pod has no node associated with it. When Kubernetes scheduler finds a node for the pod, the pod is "scheduled" for that that node. The act of scheduling is an update operation executed against the pod's `binding` subresource. KubeMod installs a webhook that listens for UPDATE operations against that subresource and that's where it triggers the logic for updating the pod's annotations.

**Comment by rohitkumar1369** [comment]
My assumption was due to these log lines :
`{"level":"info","ts":"2023-09-13 13:45:50.621Z","logger":"dragnet-webhook","msg":"Applying ModRule patch","request uid":"f100af14-2945-4061-a875-5e42f4c9ddab","namespace":"livestream-feed-service","resource":"pods/xx-58bpp","operation":"UPDATE","patch":[{"op":"add","path":"/metadata/labels/provisioning_model","value":"spot"},{"op":"add","path":"/metadata/labels/zone","value":"asia-southeast1-a"},{"op":"add","path":"/metadata/labels/node_instance_type","value":"n2-highcpu-32"},{"op":"add","path":"/metadata/labels/region","value":"asia-southeast1"}]}
{"level":"error","ts":"2023-09-13 13:45:50.627Z","logger":"podbinding-webhook","msg":"failed to inject annotation ref.kubemod.io/nodename into pod","request uid":"a320a00c-956b-4010-a75d-a7cb7c6b86d7","namespace":"xxx","resource":"pods/xx-58bpp","operation":"CREATE","podNamespace":"xxx","podName":"xx-58bpp","error":"Operation cannot be fulfilled on pods \"xx-58bpp\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\ngithub.com/kubemod/kubemod/core.(*PodBindingWebhookHandler).Handle\n\t/workspace/core/podbindingwebhook.go:95\nsigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/webhook/admission/webhook.go:135\nsigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/webhook/admission/http.go:87\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\ngithub.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2\n\t/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:76\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2462\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2916\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1966"}`

First line shows the correct patch items for the pods , but the podbinding webhook throws an error so the patch is not applied to the pod.

**Comment by rohitkumar1369** [comment]
Can i do anything to mitigate this issue with a workaround ?

**Comment by vassilvk** [comment]
So the two log items show the effect of changing the annotation.

Here's the order of what you are observing:
1. The podbinding webhook handler detects that the pod was scheduled to a node
2. It grabs the identity of the node and sticks it into an annotation on the pod
3. It then sends a request to Kubernetes to update the pod
    - That request is captured by Kubemod's dragnet webhook - the main webhook which performs Kubemod patching
    - The Kubemod webhook successfully patches the pod's manifest, but note that this is not a promise that the mutation will actually get stored. A mutating webhook only modifies the manifest of an object and returns the modified version to Kubernetes. Kubernetes is then supposed to store the mutated manifest, but this happens outside of the mutating webhook handler.
    - Up to this point, the pod itself has not been updated - only its manifest has been mutated by Kubemod and it has been returned to Kubernetes
    - Once Kubernetes gets the mutated manifest from the dragnet webhook, it attempts to store it in its etcd store
    - At that point Kubernetes' update-conflict-detection mechanism detects that someone else already modified the pod. This leads to Kubernetes responding with an error to the caller.
    - Well, that caller is the podbinding webhook
  4. The podbinding webhook reports the error and the update of the pod ultimately fails.

Bottom line is, the only thing I can suggest is implementing the retry logic as my original post.

**Comment by rohitkumar1369** [comment]
okay , let me see what i can do
